pcap_linux.go

```diff
@@ -108,15 +110,40 @@ func (h *Handle) ReadPacketData() (data []byte, ci gopacket.CaptureInfo, err err
 
 func (h *Handle) readPacketDataSyscall() (data []byte, ci gopacket.CaptureInfo, err error) {
 	b := make([]byte, h.snaplen)
-	read, _, err := syscall.Recvfrom(h.fd, b, 0)
+	oob := make([]byte, syscall.CmsgSpace(tpacketAuxdataSize))
+	n, _, _, _, err := syscall.Recvmsg(h.fd, b, oob, 0)
 	if err != nil {
-		return nil, ci, fmt.Errorf("error reading: %v", err)
+		return nil, ci, fmt.Errorf("error reading packets: %w", err)
+	}
+
+	var auxData syscall.TpacketAuxdata
+	cmsgs, err := syscall.ParseSocketControlMessage(oob)
+	if err != nil {
+		return nil, ci, fmt.Errorf("error reading socket control messages: %w", err)
+	}
+	for _, cmsg := range cmsgs {
+		if cmsg.Header.Level == syscall.SOL_PACKET && cmsg.Header.Type == syscall.PACKET_AUXDATA && cmsg.Header.Len >= tpacketAuxdataSize {
+			auxData.Vlan_tci = binary.BigEndian.Uint16(cmsg.Data[len(cmsg.Data)-5 : len(cmsg.Data)-3])
+			auxData.Vlan_tpid = binary.BigEndian.Uint16(cmsg.Data[len(cmsg.Data)-3:])
+			break
+		}
+	}
+	if auxData.Vlan_tci != 0 {
+		if auxData.Vlan_tpid == 0 {
+			auxData.Vlan_tpid = binary.BigEndian.Uint16(b[12:14])
+			binary.BigEndian.PutUint16(b[12:14], 0x8100) // set ethernet frame type to VLAN
+		}
+		aux := make([]byte, 4)
+		binary.BigEndian.PutUint16(aux[:2], auxData.Vlan_tci)
+		binary.BigEndian.PutUint16(aux[2:], auxData.Vlan_tpid)
```

Comment:
Is all of this definitely big-endian? Or is it system-dependent?

Reply:
other 802.1q (vlan tag) parsers seem to leave this as big-endian, so I felt safe leaving this as big-endian, I know over the wire, it's only ever big-endian. Here's gopacket's parser for reference https://github.com/google/gopacket/blob/master/layers/dot1q.go#L36